### DISCUSSION TITLE
Gatsby Cloud preview is very slow (3 min plus) using `gatsby-source-contentful` plugin

### DISCUSSION THREAD

**Discussion by deconstructionalism** [gatsbyjs/gatsby#31280] [Help]
## Description

We just started using Gatsby Cloud on a trial run with our organization. We host 600+ static sites with all of the content being modeled and persisted within Contentful. We recently began re-tooling our querying for site data to try and take advantage of Cloud Previews. 

The way we use Contentful is that we have a top-level content model called `Website` where each site has a unique key called `siteId`. All of the site data is joined into the `Website` content instance via content relations. For a given instance of the app, we only need the data for one site via it's `siteId`. IE, each Gatsby app instance serves previews for one site.

When our Gatsby app builds, the `gatsby-contentful-source` plugin gets the data for the _entire_ space on Contentful, ie 600+ sites and all of their content. This takes a long time for sourcing, fetching, syncing and building. For builds, this isn't so much an issue as we do not need published edits to show near-real-time changes in the published site. 

However, we do want near-real-time updates to the preview sites and it seems that the Preview builds spend a lot of time processing data, which is the bottleneck step for time efficiency. You can see that the Preview re-build to ~1 min on Gatsby Cloud for a change in one text field:

```
success Contentful: Process data (rt5zmd3ipxai-arjun-template-prototype) - 62.935s
info Creating 1051 Contentful Hero Card nodes
info Creating 744 Contentful Hero carousel nodes
info Creating 3931 Contentful Link nodes
info Creating 641 Contentful Website nodes
info Creating 591 Contentful Review Blade nodes
info Creating 1542 Contentful Service Details Card nodes
info Creating 282 Contentful Override Background nodes
info Creating 8 Contentful 60-40 Panel Blade nodes
info Creating 1 Contentful Template - Services Directory 3.0 nodes
info Creating 383 Contentful Documents and Forms nodes
info Creating 366 Contentful Online Appointment iFrame nodes
info Creating 63 Contentful Blog Category nodes
info Creating 13 Contentful Location Blade - Multiple nodes
info Creating 4978 Contentful Services Blade - Block Listing nodes
info Creating 11 Contentful Location Information nodes
info Creating 508 Contentful Gallery nodes
info Creating 475 Contentful 3-Column Text Blocks nodes
info Creating 13 Contentful Jump Link nodes
info Creating 6 Contentful Jump Link Group nodes
info Creating 81 Contentful Accordion Blade nodes
info Creating 791 Contentful Accordion nodes
info Creating 10342 Contentful Staff nodes
info Creating 2271 Contentful Staff Blade nodes
info Creating 6 Contentful Staff Group nodes
info Creating 1 Contentful Staff Accreditation nodes
info Creating 68 Contentful Image Slider nodes
info Creating 1 Contentful Services Blade - Detailed Listing nodes
info Creating 1472 Contentful Service Details nodes
info Creating 890 Contentful Rich Text Blob nodes
info Creating 93 Contentful Blog Article nodes
info Creating 23 Contentful Pencil Banner nodes
info Creating 1210 Contentful Secondary Hero nodes
info Creating 163 Contentful Key Category Menu nodes
info Creating 19 Contentful Item Label nodes
info Creating 171 Contentful Key Category Menu Group nodes
info Creating 616 Contentful Website Header nodes
info Creating 654 Contentful Location Information Blade nodes
info Creating 1910 Contentful Promo Banner nodes
info Creating 1765 Contentful Review Card nodes
info Creating 609 Contentful Website Footer nodes
info Creating 784 Contentful Social Links Container nodes
info Creating 2169 Contentful 60-40 Panel nodes
info Creating 585 Contentful 5-Card Category Cards Blade nodes
info Creating 1067 Contentful 5-Card Category Card nodes
info Creating 488 Contentful 3-Tile Promotion Cards Blade nodes
info Creating 892 Contentful 3-Tile nodes
info Creating 134 Contentful Video nodes
info Creating 19685 Contentful Image nodes
info Creating 475 Contentful Custom Code nodes
info Creating 5539 Contentful Website Page nodes
info Creating 24 Contentful Story Blade nodes
info Creating 23 Contentful Blog Blade nodes
info Creating 620 Contentful Location Container nodes
info Creating 597 Contentful FormKeep Form nodes
info Creating 25339 Contentful asset nodes
```

- Is there a way to get previews to be faster without completing re-factoring our content model? 
- Or otherwise, how would you advise Contentful-sourced apps with hundreds of sites to take advantage of Preview builds?  

### Steps to reproduce

This is more a question content architecture in relation to Gatsby Cloud architecture. We can't really advise on how to repro this unless you have access to our deeply nested content model.

### Expected result

Near-instant preview build updates.

### Actual result

Preview build updates took over a minute. 

### Environment

```
  System:
    OS: macOS 11.2.1
    CPU: (4) x64 Intel(R) Core(TM) i7-7660U CPU @ 2.50GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.1 - ~/.nvm/versions/node/v14.16.1/bin/node
    Yarn: 1.5.1 - /usr/local/bin/yarn
    npm: 6.14.12 - ~/.nvm/versions/node/v14.16.1/bin/npm
  Languages:
    Python: 3.7.9 - /Users/arjunray/.pyenv/shims/python
  Browsers:
    Edge: 90.0.818.51
    Firefox: 62.0
    Safari: 14.0.3
  npmPackages:
    gatsby: ^3.4.0 => 3.4.0
    gatsby-plugin-image: ^1.4.0 => 1.4.0
    gatsby-plugin-sharp: ^3.4.0 => 3.4.0
    gatsby-source-contentful: 5.5.0-next.0 => 5.5.0-next.0
    gatsby-transformer-remark: ^4.0.0 => 4.1.0
    gatsby-transformer-sharp: ^3.0.0 => 3.4.0
```

**Comment by deconstructionalism** [comment]
Also, I've been looking at the config for the `gatsby-source-contentful` plugin and I don't see any way to segment the data such that only one `siteId`'s data is fetched. Am I missing something or is this not actually possible to do with the plugin?

**Comment by KyleAMathews** [comment]
Huh wow. For fun, how many nodes are being created?

Tagging @axe312ger.

He would know if contentful's API allows for filtering your content to just the content for a single site. If so, seem pretty reasonable to support that in the source plugin.

**Comment by deconstructionalism** [comment]
@KyleAMathews 

So this line here could be modified in order to sync a specific content instead of all content:

https://github.com/gatsbyjs/gatsby/blob/4ecf00a68b76d758f45020a8500c73c4577d9de1/packages/gatsby-source-contentful/src/fetch.js#L230


by using the `content_type` option as documented [here](https://contentful.github.io/contentful.js/contentful/8.3.3/ContentfulClientAPI.html#.sync).

I'm going to try some performance measurements using the data in my space with and without setting the `content_type` to `site` (the Website content type in our space).

**Reply by deconstructionalism** [reply]
So, simply passing `content_type: 'site'` dramatically improves the duration of the `sync` operation. It takes only a split second to get the Contentful data for all sites  (as opposed to a few minutes without `content_type` set) and it seems to have all links resolved, so the raw data is highly nested with what I assume is all related data intact. 

However, the following logic in `gatsby-source-contentful` expects the shape of the `sync` data to include entity data for each content model on Contentful, ie it consumes the flat list of entity collections for each entity type and does not recurse through individual entity data to pull out related content. 

_roughly:_
```js
// what the plugin expects

{
  Dogs: [dog1, dog2, ...restOfDogs],
  Owners: [owner1, owner2, ...restOfOwners],
  Houses: [houses1, houses2, ...restOfHouses],
  Sites: [site1, site2, ...restOfSites]
}

// in this dummy example, each site contains related houses, owners and dogs:

{
  name: site1,
  houses: [
    {
      name: house1,
      dogs: [
        { 
          name: dog1
        },
        { 
          name: dog2
        }
      ]
    }
  ]
}
```

_vs what it gets via `sync` with `content_type: 'site'`_:
```js
// what the plugin gets

{
  Sites: [site1, site2, ...restOfSites]
}
```

With `content_type: 'site'`, only the `site` content model returned by `sync`, the plugin fails to consume the deeply nested data within the `site` Entities. What ends up in Gatsby are named nodes for each content type in the Contentful space, but only the `site` nodes have data, and the data is shallow, ie all the nested relations inside the raw sync data for sites are lost.

**Reply by deconstructionalism** [reply]
If you all think this is a missing feature, it might make sense to have someone who's worked closely on that plugin take a crack at it first. Otherwise, I can try and put in a PR for it at some point in the future if we are still considering Gatsby Cloud as a solution

**Reply by axe312ger** [reply]
Contentful recommends you to use one space per website/project.

When reducing the content_type to a single one, you risk unresolved links from content types that are not synced. (As far as I know)


So if you need to query only a fraction of your space:

1. remove all the other content form the space, create multiple
2. migrate your code to directly use the GraphQL API from Contentful. That way you can only query what you need. Unfortunately I can't guide you how to set it up, I never did it for myself.

**Reply by deconstructionalism** [reply]
This wouldn't really work for us because all of the sites share the same exact content model, as they all use the same gatsby app to render static sites. Our app and our content model are tightly coupled so you can imagine that we may iterate our content model and our app to match it, and then be stuck with a nightmare of trying to get all the content models in all of the 600+ spaces consistent. If Contentful allowed spaces to share content models, this wouldn't be an issue but they currently do not

**Reply by deconstructionalism** [reply]
The old version of our app did query using the generic `gatsby-source-graphql` plugin. It looks like they were not restricting the query by `siteId` so perhaps we could reduce query times that way

**Reply by deconstructionalism** [reply]
@axe312ger  @KyleAMathews 

Also, is it correct that incremental builds and preview builds for Contentful sourced apps are only available on Gatsby Cloud if we use the `gatsby-source-contentful` plugin?

**Reply by deconstructionalism** [reply]
do you need to use `gatsby-source-contentful` to have access to Gatsby Cloud previews and incremental builds?

**Reply by KyleAMathews** [reply]
No it works with any source plugin. You could e.g. try publishing your version of gatsby-source-contentful with the change to NPM and deploy it to Gatsby Cloud to see how it works.

**Reply by axe312ger** [reply]
@deconstructionalism

Using a single space for 600 websites is far from the intended use. Contentful wants you to use one space per project. That is how the (sync) API works as well. We can't filter entries, we just get all of them.

That means even when we would want to support that uncommon use, we can not.

Please split up your space into a space per website/project:

You can use https://www.contentful.com/developers/docs/tutorials/cli/scripting-migrations/ and https://www.contentful.com/faq/content-migrations/ to keep the content model in sync.

You can also reach out to https://www.contentful.com/support/. They might be able to help you with splitting up the space.
